src/Database/LSMTree/Internal/RefCount.hs

```diff
@@ -0,0 +1,115 @@
+unsafeMkRefCounterN :: PrimMonad m => RefCount -> Maybe (m ()) -> m (RefCounter m)
+unsafeMkRefCounterN !n finaliser = mkRefCounterN n finaliser >>= \case
+    Nothing -> error "unsafeMkRefCounterN: n < 1"
+    Just rc -> pure rc
+
+{-# SPECIALISE mkRefCounterN :: RefCount -> Maybe (IO ()) -> IO (Maybe (RefCounter IO)) #-}
+-- | Make a reference counter with initial value @n@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+--
+-- Returns 'Nothing' if @n < 1@, and 'Just' otherwise.
+mkRefCounterN :: PrimMonad m => RefCount -> Maybe (m ()) -> m (Maybe (RefCounter m))
+mkRefCounterN !n finaliser
+  | n < RefCount 1 = pure Nothing
+  | otherwise = do
+    -- evaluate the finaliser a little bit before we store it
+    let !() = case finaliser of
+                Nothing       -> ()
+                Just !_action -> ()
+    countVar <- newMutVar n
+    finaliserVar <- newMutVar finaliser
+    pure $! Just $! RefCounter{countVar, finaliserVar}
+
+{-# SPECIALISE mkRefCounter1 :: Maybe (IO ()) -> IO (RefCounter IO) #-}
+-- | Make a reference counter with initial value @1@. An optional finaliser is
+-- run when the reference counter reaches @0@.
+mkRefCounter1 :: PrimMonad m => Maybe (m ()) -> m (RefCounter m)
+mkRefCounter1 finaliser = fromJust <$> mkRefCounterN (RefCount 1) finaliser
+
+data RefCountError = RefCountAlreadyZero
+  deriving stock (Show, Eq)
+  deriving anyclass Exception
+
+{-# SPECIALISE addReference :: RefCounter IO -> IO () #-}
+-- | Increase the reference counter by one.
+--
+-- If the reference counter was already at @0@, a 'RefCountError' error is
+-- thrown.
+addReference :: (PrimMonad m, MonadThrow m) => RefCounter m -> m ()
+addReference RefCounter{countVar} = do
+    success <- atomicModifyMutVar' countVar $ \c ->
+      if c == RefCount 0 then (c, False)
+                         else (incRefCount c, True)
+    unless success $ throwIO RefCountAlreadyZero
+
+{-# SPECIALISE readRefCount :: RefCounter IO -> IO RefCount #-}
+readRefCount :: PrimMonad m => RefCounter m -> m RefCount
+readRefCount RefCounter{countVar} = readMutVar countVar
+
+{-# SPECIALISE removeReference :: RefCounter IO -> IO () #-}
+-- | Decrease the reference counter by one.
+--
+-- When the reference counter reaches zero, the finaliser is run promptly.
+--
+-- If the reference counter was already at @0@, then this function is
+-- idempotent. The finaliser will not run again.
+removeReference :: PrimMonad m => RefCounter m -> m ()
+removeReference rc@RefCounter{countVar} = do
+    may <- atomicModifyMutVar' countVar $ \c ->
+      if c == RefCount 0 then (c, Nothing)
+                         else let c' = decRefCount c in (c', Just (c == RefCount 1))
+    case may of
+      Nothing -> pure ()
+      Just b  -> when b $ finalise rc
+
+{-# SPECIALISE replaceFinaliser :: RefCounter IO -> Maybe (IO ()) -> IO () #-}
+replaceFinaliser :: PrimMonad m => RefCounter m -> Maybe (m ()) -> m ()
```

Comment:
Why do we need this?

Reply:
Note that #353 does not use `replaceFinaliser`. Unless you know that we need it, leave it out until such a time as we discover that it's necessary. I can't currently think of a use case.

Reply:
It's a slightly awkward design, I admit. We do use it in #353, specifically in lines https://github.com/IntersectMBO/lsm-tree/pull/353/files#diff-c5f8e29426788087d6648fdbf6404e7fbb9c5c710e22f88bf096f0951302e6f3R185-R195. It's done this way because there is circularity: a `Run` holds a `RefCounter`, and then the finaliser inside the `RefCounter` is a function that is called on a `Run`

Reply:
In any case, I'm probably getting rid of the finaliser, so this problem probably disappears

Reply:
> Why do we need this?

We don't *need* this -- it's an attempt at a unified approach to reference counting + recording what action should be performed once a reference count reaches 0. One example of a use case for the finaliser is `Run.close`

Reply:
We decided to keep the finaliser